deepmd/dpmodel/model/spin_model.py

```diff
@@ -0,0 +1,384 @@
+        return extended_nlist
+
+    @staticmethod
+    def concat_switch_virtual(extended_tensor, extended_tensor_virtual, nloc: int):
+        nframes, nall = extended_tensor.shape[:2]
+        out_shape = list(extended_tensor.shape)
+        out_shape[1] *= 2
+        extended_tensor_updated = np.zeros(
+            out_shape,
+            dtype=extended_tensor.dtype,
+        )
+        extended_tensor_updated[:, :nloc] = extended_tensor[:, :nloc]
+        extended_tensor_updated[:, nloc : nloc + nloc] = extended_tensor_virtual[
+            :, :nloc
+        ]
+        extended_tensor_updated[:, nloc + nloc : nloc + nall] = extended_tensor[
+            :, nloc:
+        ]
+        extended_tensor_updated[:, nloc + nall :] = extended_tensor_virtual[:, nloc:]
+        return extended_tensor_updated.reshape(out_shape)
+
+    def get_type_map(self) -> List[str]:
+        """Get the type map."""
+        tmap = self.backbone_model.get_type_map()
+        ntypes = len(tmap) // 2  # ignore the virtual type
+        return tmap[:ntypes]
+
+    def get_rcut(self):
+        """Get the cut-off radius."""
+        return self.backbone_model.get_rcut()
+
+    def get_dim_fparam(self):
+        """Get the number (dimension) of frame parameters of this atomic model."""
+        return self.backbone_model.get_dim_fparam()
+
+    def get_dim_aparam(self):
+        """Get the number (dimension) of atomic parameters of this atomic model."""
+        return self.backbone_model.get_dim_aparam()
+
+    def get_sel_type(self) -> List[int]:
+        """Get the selected atom types of this model.
+        Only atoms with selected atom types have atomic contribution
+        to the result of the model.
+        If returning an empty list, all atom types are selected.
+        """
+        return self.backbone_model.get_sel_type()
+
+    def is_aparam_nall(self) -> bool:
+        """Check whether the shape of atomic parameters is (nframes, nall, ndim).
+        If False, the shape is (nframes, nloc, ndim).
+        """
+        return self.backbone_model.is_aparam_nall()
+
+    def model_output_type(self) -> str:
+        """Get the output type for the model."""
+        return self.backbone_model.model_output_type()
+
+    def get_model_def_script(self) -> str:
+        """Get the model definition script."""
+        return self.backbone_model.get_model_def_script()
+
+    def get_nnei(self) -> int:
+        """Returns the total number of selected neighboring atoms in the cut-off radius."""
+        # for C++ interface
+        return self.backbone_model.get_nnei() // 2  # ignore the virtual selected
+
+    def get_nsel(self) -> int:
+        """Returns the total number of selected neighboring atoms in the cut-off radius."""
+        return self.backbone_model.get_nsel() // 2  # ignore the virtual selected
+
+    @staticmethod
+    def has_spin() -> bool:
+        """Returns whether it has spin input and output."""
+        return True
```

Comment:
To be the same as pt model, which needs jit export. So it doesn't use @property here.